memory_model/src/guest_memory.rs

```diff
@@ -138,6 +146,112 @@ impl GuestMemory {
         self.regions.len()
     }
 
+    fn get_pfn(num: u64) -> u64 {
+        num & 0x7FFFFFFFFFFFFF
+    }
+
+    /// return a mapping from guest physical page numbers to host physical page numbers
+    /// for the given virtual address range
+    fn read_pagemap_addr_range(page_i_base: usize, page_size: u64, addr: u64, size: u64)
+    -> BTreeMap<u64, usize> {
+        const PM_ENTRY_SIZE:u64 = 8;
+        let path = format!("/proc/{}/pagemap", std::process::id());
+        let offset = addr/page_size*PM_ENTRY_SIZE;
+
+        let mut pagemap = std::fs::File::open(&path).unwrap();
+        pagemap.seek(SeekFrom::Start(offset)).unwrap_or_default();
+
+        let num_pages = (size / page_size) as usize;
+        let mut buf = [0 as u8; 8];
+        let mut pfns = BTreeMap::new();
+        for page_i in 0..num_pages {
+            pagemap.read_exact(&mut buf).err();
+            let entry = u64::from_le_bytes(buf);
+            // check if the page is present
+            if get_bit(entry, 63) == 1 {
+                pfns.insert(GuestMemory::get_pfn(entry), page_i_base + page_i);
+            }
+        }
+        pfns
+    }
+
+    /// return a mapping from guest physical page numbers to host physical page numbers
+    /// for the entire guest memory
+    pub fn get_pagemap(&self) -> BTreeMap<u64, usize> {
+        let page_size = unsafe { libc::sysconf(libc::_SC_PAGESIZE) } as u64;
+        let mut pfns = BTreeMap::new();
+        let mut page_i_base = 0 as usize;
+        for region in self.regions.iter() {
+            pfns.append(
+                &mut GuestMemory::read_pagemap_addr_range(page_i_base,
+                                                          page_size,
+                                                          region.mapping.as_ptr() as u64,
+                                                          region.mapping.size() as u64)
+            );
+            page_i_base += region.mapping.size() / page_size as usize;
+        }
+        pfns
+    }
+
+    /// Write all initialized guest memory pages to the provided writer
+    /// Here being initialized means being present in physical RAM
+    /// The byte stream being written out consists of a sequence of
+    /// guest physical frame (i.e. page) number (gpfn) in little endian (8 Bytes)
+    /// immediately followed by the corresponding guest page (4096 Bytes)
+    pub fn dump_init<F>(&self, writer: &mut F) -> Result<()>
```

Comment:
perhaps the more verbose but clearer `dump_initialized_memory`, or just `dump_memory` (that you're only dumping initialized memory is an optimization)